### PR TITLE
Add support for System.IO.Directory

### PR DESCRIPTION
**Pull request by morali**
## Description

- Implements native support for Directory API for System.IO.FileSystem.

## Motivation and Context

- Following nanoframework/System.IO.FileSystem#17.

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (work on Unit Tests, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+HRESULT Library_nf_sys_io_filesystem_System_IO_Directory::ExistsNative___STATIC__BOOLEAN__STRING(
+    CLR_RT_StackFrame &stack)
+{
+    NANOCLR_HEADER();
+
+    const char *folderPath = stack.Arg0().RecoverString();
+
+    bool exists = false;
+    FRESULT operationResult;
+    FAULT_ON_NULL(folderPath);
```
Comment (josesimoes) [inline src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp:21]: These should be `FAULT_ON_NULL_ARG` which will throw the ArgumentNull exception.
This one is for NullReference, which is not the intended exception.

File: src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+    NANOCLR_NOCLEANUP();
+}
+
+HRESULT Library_nf_sys_io_filesystem_System_IO_Directory::MoveNative___STATIC__VOID__STRING__STRING(
+    CLR_RT_StackFrame &stack)
+{
+    NANOCLR_HEADER();
+    {
+        const char *filePathSrc = stack.Arg0().RecoverString();
+        FAULT_ON_NULL(filePathSrc);
```
Comment (josesimoes) [inline src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp:47]: same here

File: src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+        const char *filePathDest = stack.Arg1().RecoverString();
+        FAULT_ON_NULL(filePathDest);
```
Comment (josesimoes) [inline src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp:49]: same here

File: src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+    }
+    NANOCLR_NOCLEANUP();
+}
+
+HRESULT Library_nf_sys_io_filesystem_System_IO_Directory::DeleteNative___STATIC__VOID__STRING(CLR_RT_StackFrame &stack)
+{
+    NANOCLR_HEADER();
+    {
+        const char *folderPath = stack.Arg0().RecoverString();
+        FAULT_ON_NULL(folderPath);
```
Comment (josesimoes) [inline src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp:73]: same here

File: src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+    }
+    NANOCLR_NOCLEANUP();
+}
+
+HRESULT Library_nf_sys_io_filesystem_System_IO_Directory::CreateNative___STATIC__VOID__STRING(CLR_RT_StackFrame &stack)
+{
+    NANOCLR_HEADER();
+    {
+        const char *folderPath = stack.Arg0().RecoverString();
+        FAULT_ON_NULL(folderPath);
```
Comment (josesimoes) [inline src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp:97]: same here

File: src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+        CLR_RT_HeapBlock *filePathEntry;
+        const char *folderPath = stack.Arg0().RecoverString();
+        CLR_RT_HeapBlock &filePaths = stack.PushValue();
+
+        uint16_t fileCount = 0;
+        DIR currentDirectory;
+        FRESULT operationResult;
+        static FILINFO fileInfo;
+
+        FAULT_ON_NULL(folderPath);
```
Comment (josesimoes) [inline src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp:130]: same here

File: src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+        CLR_RT_HeapBlock *folderPathEntry;
+        const char *folderPath = stack.Arg0().RecoverString();
+        CLR_RT_HeapBlock &folderPaths = stack.PushValue();
+
+        uint16_t directoryCount = 0;
+        DIR currentDirectory;
+        FRESULT operationResult;
+        static FILINFO fileInfo;
+
+        FAULT_ON_NULL(folderPath);
```
Comment (josesimoes) [inline src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp:260]: same here

File: src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+    CLR_RT_StackFrame &stack)
+{
+    NANOCLR_HEADER();
+    {
+        SYSTEMTIME fileInfoTime;
+        CLR_RT_TypeDescriptor dtType;
+        FILINFO fileInfo;
+
+        const char *folderPath = stack.Arg0().RecoverString();
+        FAULT_ON_NULL(folderPath);
```
Comment (josesimoes) [inline src/System.IO.FileSystem/nf_sys_io_filesystem_System_IO_Directory.cpp:412]: same here

**Review by josesimoes — CHANGES_REQUESTED** [review]
Overall looking good. Just replace those exception throwers and we are good.

Nicely done! 👍🏻

**Comment by morali** [comment]
> 
> 
> Overall looking good. Just replace those exception throwers and we are good.
> 
> Nicely done! 👍🏻

Thank you, the exceptions throwers were replaced.

**Review by josesimoes — APPROVED** [review]
LGTM!

**Pull request merged**